src/metrics/tagfiltertree/tag_filter_tree.go

```diff
@@ -0,0 +1,317 @@
+package tagfiltertree
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	_matchall  = "*"
+	_matchNone = "!*"
+)
+
+// Tag represents a tag with name, value and variable.
+type Tag struct {
+	Name string
+	Val  string
+	Var  string
```

Comment:
where is the usage of Var? I don't find any

Reply:
Var will be used in Namespace Attribution to support capture-groups like functionality in Regex.

Reply:
so it will not be used in query matching? nit: add some comments to make the code more readable